corehq/ex-submodules/dimagi/utils/couch/__init__.py

```diff
@@ -222,18 +232,20 @@ def get_locked_obj(cls, *args, **kwargs):
                     release_lock(lock, degrade_gracefully)
                     return LockManager(None, None)
         except:
+            exc = sys.exc_info()
             release_lock(lock, degrade_gracefully)
-            raise
-        else:
-            if _id:
-                return LockManager(obj, lock)
-            else:
-                obj_lock = cls.get_obj_lock(obj)
-                obj_lock = acquire_lock(obj_lock, degrade_gracefully)
+            six.reraise(*exc)
+        if _id:
+            return LockManager(obj, lock)
+        try:
+            obj_lock = cls.get_obj_lock(obj)
+            obj_lock = acquire_lock(obj_lock, degrade_gracefully, blocking=block)
+            if block:
                 # Refresh the object in case another thread has updated it
                 obj = cls.get_latest_obj(obj)
-                release_lock(lock, degrade_gracefully)
-                return LockManager(obj, obj_lock)
+            return LockManager(obj, obj_lock)
+        finally:
+            release_lock(lock, degrade_gracefully)
```

Comment:
this code always confuses me, too many locks and alternate code paths!

Reply:
I agree!